### PR TITLE
ref(py3): Use compat izip, iteritems, itervalues, and types

### PR DESCRIPTION
**Pull request by EvanPurkhiser**
Maybe we swap six in at some point

### CODE REVIEW

File: tests/test_cluster.py [hunk 1]
```diff
@@ -185,7 +188,7 @@ def test_fanout_api(cluster):
 def test_fanout_key_target(cluster):
     with cluster.fanout() as client:
         c = client.target_key('foo')
-        c.set('foo', str(42))
+        c.set('foo', bytes_type(42))
```
Comment (wedamija) [inline tests/test_cluster.py:191]: Are you sure they want bytes here?  When using `set` the values will be string. It looks to me like they want `42` to be a string like `"42"`.

File: tests/test_cluster.py [hunk 2]
```diff
@@ -211,7 +214,7 @@ def test_emulated_batch_apis(cluster):
     assert promise.value is None
     with cluster.map() as map_client:
         promise = map_client.mget(['key:%s' % x for x in range(10)])
-    assert promise.value == [str(x) for x in range(10)]
+    assert promise.value == map(bytes_type, range(10))
```
Comment (wedamija) [inline tests/test_cluster.py:217]: Same here.

File: rb/utils.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
+from __future__ import absolute_import
+
+import sys
+
+PY2 = sys.version_info[0] == 2
+
+if PY2:
+    integer_types = (int, long)
+
+    text_type = type(u"")
```
Comment (wedamija) [inline rb/utils.py:10]: `unicode`

File: rb/utils.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+    bytes_type = type("")
```
Comment (wedamija) [inline rb/utils.py:12]: `str`

File: rb/utils.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+    _iteritems = "iteritems"
+    _itervalues = "itervalues"
+
+    from itertools import izip
+else:
+    integer_types = (int,)
+
+    text_type = type("")
```
Comment (wedamija) [inline rb/utils.py:21]: `str`

File: rb/utils.py [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+
+    bytes_type = type(b"")
```
Comment (wedamija) [inline rb/utils.py:23]: `bytes`

File: rb/router.py [hunk 7]
```diff
@@ -132,8 +133,8 @@ def __init__(self, cluster):
         assert_gapless_hosts(self.cluster.hosts)
 
     def get_host_for_key(self, key):
-        if isinstance(key, unicode):
+        if isinstance(key, text_type):
             k = key.encode('utf-8')
         else:
-            k = str(key)
+            k = bytes_type(key)
```
Comment (wedamija) [inline rb/router.py:139]: Will `key` always be a string or bytes? If it's an integer then using `bytes` here in python 3 will be bad. I feel like the intent of the `str` here was "Make sure this is a string", in terms of converting ints to a string (which effectively means "ascii encoded bytes", I suppose).

File: tests/test_cluster.py [hunk 8]
```diff
@@ -271,7 +274,7 @@ def test_execute_commands(cluster):
 def test_reconnect(cluster):
     with cluster.map() as client:
         for x in range(10):
-            client.set(str(x), str(x))
+            client.set(bytes_type(x), bytes_type(x))
```
Comment (wedamija) [inline tests/test_cluster.py:277]: Same here

File: tests/test_cluster.py [hunk 9]
```diff
@@ -280,8 +283,8 @@ def test_reconnect(cluster):
 
     with cluster.map() as client:
         rv = Promise.all([
-            client.get(str(x))
+            client.get(bytes_type(x))
             for x in range(10)
         ])
 
-    assert rv.value == list(map(str, range(10)))
+    assert rv.value == list(map(bytes_type, range(10)))
```
Comment (wedamija) [inline tests/test_cluster.py:290]: And here

File: tests/test_poll.py [hunk 10]
```diff
@@ -15,4 +16,4 @@ def test_simple_api(cluster, poll, monkeypatch):
             map_client.set('key:%s' % x, x)
 
     for x in range(10):
-        assert client.get('key:%d' % x) == str(x)
+        assert client.get('key:%d' % x) == bytes_type(x)
```
Comment (wedamija) [inline tests/test_poll.py:19]: Probably same here too

File: tests/test_cluster.py (hunk 1 above)
Comment (EvanPurkhiser) [reply tests/test_cluster.py:191]: I copied this over from dcramers stuff fwiw

Comment (wedamija) [reply tests/test_cluster.py:191]: In Py2:

```
In [9]: str(42)
Out[9]: '42'
```

In Py3
```
>>> bytes(42)
b'\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00'
```

Comment (EvanPurkhiser) [reply tests/test_cluster.py:191]: lol you're right

File: rb/router.py (hunk 7 above)
Comment (EvanPurkhiser) [reply rb/router.py:141]: I've guarded against integer types and coerced them into strings now. I left the byte conversion fallback.

File: tests/test_poll.py (hunk 10 above)
Comment (EvanPurkhiser) [reply tests/test_poll.py:19]: I think you're probably right, lets see what happens in py3 land.

Going to leave this as is for now since tests pass

Comment (wedamija) [reply tests/test_poll.py:19]: I think we should fix this now, we're passing ints to `bytes` in python 3, which will generate a giant bytestring and fail. `text_type` will work fine here.

Comment (EvanPurkhiser) [reply tests/test_poll.py:19]: oh doy, you're right

File: rb/utils.py [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+    _iteritems = "items"
+    _itervalues = "values"
+
+
+def iteritems(d, **kw):
+    return iter(getattr(d, _iteritems)(**kw))
+
+
+def itervalues(d, **kw):
+    return iter(getattr(d, _itervalues)(**kw))
```
Comment (BYK) [inline rb/utils.py:36]: The following would probably be more performant (if these are called frequently in hot spots):

```py
if PY2:
    ...
    def iteritems(d, **kw):
        return iter(d.iteritems(**kw))
    
    def itervalues(d, **kw):
        return iter(d.itervalues(**kw))
else:
    ...
    def iteritems(d, **kw):
        return iter(d.items(**kw))
    
    def itervalues(d, **kw):
        return iter(d.values(**kw))
```

Comment (EvanPurkhiser) [reply rb/utils.py:36]: Yeah good call, will swap it over to that

Comment (EvanPurkhiser) [reply rb/utils.py:36]: https://github.com/getsentry/rb/pull/43

**Pull request merged**